prefect_email/credentials.py

```diff
@@ -95,6 +96,7 @@ class EmailServerCredentials(Block):
     smtp_server: Optional[Union[str, SMTPServer]] = SMTPServer.GMAIL
     smtp_type: Optional[Union[str, SMTPType]] = SMTPType.SSL
     smtp_port: Optional[int] = None
+    context: Optional[object] = None
```

Comment:
```suggestion
    context: Optional[ssl.SSLContext] = None
```

Reply:
I initially did do this but pytest/pydantic complained with:

no validator found for <class 'ssl.SSLContext'>, see `arbitrary_types_allowed` in Config

I then added a class Config: arbitary_types_allowed = True etc to fix and then got this:

ValueError: Value not declarable with JSON Schema, field: name='context' type=Optional[SSLContext] required=False default=None

which means creating a custom validator and also an encoder to deal with the SSLContext. I wasn't sure if that much boilerplate / extra code just to pass a test was warranted. If you want me to, I will do it.

Reply:
Thanks for the detailed explanation. I would advocate for using SSLContext as the type annotation so users know exactly what kind of context and also better editor experience!

Similar to https://github.com/PrefectHQ/prefect-aws/blob/main/prefect_aws/ecs.py#L423-L434 I think.

Reply:
Since this block will be configurable via the UI, I think we may need to take a slightly different approach with providing a context. It looks like `SSLContext` [can be passed a protocol](https://docs.python.org/3/library/ssl.html#ssl.SSLContext) when creating a new context, so I think we could have a `protocol` attribute that is an enum that accepts any of the protocols that `SSLContext` accepts. `context` then could be a property on this class so that it's usable in `get_server`, or we could create a context on each invocation of `get_server`

Reply:
I have been at it nearly all day now trying to make pydantic happy and nothing has worked. There is an open issue in pydantic related to [this](https://github.com/pydantic/pydantic/issues/4079)

If the definition of context is  

context: Optional[Type[ssl.SSLContext]] 

then the tests at least pass, but the schema Pydantic creates is not really ideal (no real info on 'context') and also I had to upgrade pydantic to 1.10.2 to get this to work...  And I can't get pydantic to modify the schema generated for EmailServerCredentials for some reason (using __modify_schema__()). I also tried making a json_encoder etc... no dice. 

What you're suggesting makes sense, but it sounds like an extra feature and shouldn't be rolled into what I'm doing right now, I think. I have to call SSLContext methods to create my desired context and having Prefect handle those responsibilities doesn't make sense.

Given the pydantic issue, I think the efficient thing to do here is 1) mandate pydantic 1.10 is used and 2) use Type[] in the definition of 'context' (and rename context to ssl_context)

Reply:
If you need a customized SSLContext, another option is to add an optional parameter `ssl_context` parameter to `get_server`. That way you won't have to fight Pydantic, we don't have to worry about the impacts of adding an `ssl_context` attribute on our UI, and you can use a custom `SSLContext` object in your flows.

Reply:
If we do expose protocol on credentials, it seems ssl has built-in constants:
`context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)`

But, having a `context` kwarg in get_server is easy, just not as portable as inside the block.

Also wanted to say, thank you and I really appreciate the thought and effort you put in!

Reply:
oh wait, if you are just worried about documenting the field properly, I realize now that we could use pydantic's Field() and hopefully modify how it is documented. I will try that.

Reply:
I think Alex is concerned about whether the inputs can be serialized through the [Prefect UI](https://orion-docs.prefect.io/ui/blocks/).

Reply:
> If you need a customized SSLContext, another option is to add an optional parameter `ssl_context` parameter to `get_server`. That way you won't have to fight Pydantic, we don't have to worry about the impacts of adding an `ssl_context` attribute on our UI, and you can use a custom `SSLContext` object in your flows.

yeah, that makes sense as well. Adding another attribute looked like the way to go...and now I know pydantic a heck of a lot better.  Let me know what you want to do and I will get it working.

Reply:
I like the idea of protocol as a keyword argument in EmailServerCredentials.
```
class EmailServerCredentials(Block):
    """
    Block used to manage generic email server authentication.
    It is recommended you use a
    [Google App Password](https://support.google.com/accounts/answer/185833)
    if you use Gmail.

    Attributes:
        ...
        protocol: Protocol used to initialize SSLContext.

     ...
     protocol: Optional[ssl._SSLMethod] = None
```

Reply:
I have now changed things so that an optional SSLContext can be passed to get_server(). It makes more sense (if blocks are stored, how is that SSLContext going to be stored? pickle? it is too heavy to do that perhaps), avoids all the problems with pydantic, and less code changes are made to the original file. I've modified / added tests as well
So I think I'm done

I will let you all figure out what to do with 'protocol' - that is a separate change that impacts the UI.

Reply:
I think protocol directly relates to SSLContext since to instantiate SSLContext, you need to pass a protocol right? 

If that is correct, what we were thinking was we would let users pass in the corresponding protocol, and we instantiate the SSLContext internally.

Reply:
No, you do not need to pass a protocol. The default is PROTOCOL_TLS and one would override that if you needed something out of the ordinary (e.g. SSLv2 and SSLv3 are discouraged).  But I see your point so I'll add in ssl_protocol.

Reply:
Maybe we can default to `protocol: str = PROTOCOL_TLS`

Reply:
But 3.10 has overridden 3.8 in what the default is in SSLProtocol(protocol)...  and 3.10 deprecates the ability to not specify a protocol... and 3.11+ will need one of the 2 different PROTOCOL_TLS_* values.  And create_default_context() sets up other things that might not be handled by an explicitly created SSLContext:

The settings are chosen by the [ssl](https://docs.python.org/3/library/ssl.html#module-ssl) module, and usually represent a higher security level than when calling the [SSLContext](https://docs.python.org/3/library/ssl.html#ssl.SSLContext) constructor directly.

Fighting Python is never a great idea. Best to handle the most straightforward stuff while leaving enough leeway for hooks to easily alter behavior. My philosophy here is that the SSLContext is the more important thing (directly tied to the SMTP server) and protocol is merely an aspect of that - look at all the other ways you can customize a SSLContext. You don't want to have embed all that into settings of EmailServerCredentials.

Reply:
I thought of another way of explaining my position - 90+% of Prefect's users will not be concerned with security concerns b/c the defaults are sufficient (better than they could config themselves) and are handled by Python + the OS layer (ssl libs etc). So create_default_context() is the proper way to do things. Prefect should not concern itself with low level security details.

But for situations like mine, Prefect needs to enable my code to set security as it needs to - and my code does not and should not trust Prefect code at all (no offense). Thus my position of wanting Prefect to be as dumb as possible - it does the defaults as established by Python/OS or it lets my code configure things and just passes them along to the relevant Python libs it is wrapping - no magic help or trying to be clever... Prefect does not want to get into security related matters at all - and setting up 'protocol' to be able to be manipulated by the user is to start opening that door - otherwise someone else will ask how they can fiddle with other aspects of the custom SSLContext Prefect makes and then it never ends.   Does this help?  What I'm getting at are the larger, more abstract design concerns and security related ones are always difficult to get right.